### PR TITLE
[SPARK-16411][SQL][STREAMING] Add textFile to Structured Streaming.

### PR DESCRIPTION
**Pull request by ScrapCodes**
## What changes were proposed in this pull request?

Adds the textFile API which exists in DataFrameReader and serves same purpose.
## How was this patch tested?

Added corresponding testcase.

### CODE REVIEW

File: sql/core/src/main/scala/org/apache/spark/sql/streaming/DataStreamReader.scala [hunk 1]
```diff
@@ -281,6 +281,31 @@ final class DataStreamReader private[sql](sparkSession: SparkSession) extends Lo
   @Experimental
   def text(path: String): DataFrame = format("text").load(path)
 
+  /**
+   * Loads text files and returns a [[Dataset]] of String. The underlying schema of the Dataset
```
Comment (jaceklaskowski) [inline sql/core/src/main/scala/org/apache/spark/sql/streaming/DataStreamReader.scala]: a text file?

File: sql/core/src/main/scala/org/apache/spark/sql/streaming/DataStreamReader.scala [hunk 2]
```diff
@@ -281,6 +281,31 @@ final class DataStreamReader private[sql](sparkSession: SparkSession) extends Lo
...
+   * contains a single string column named "value".
+   *
+   * If the directory structure of the text files contains partitioning information, those are
+   * ignored in the resulting Dataset. To include partitioning information as columns, use `text`.
+   *
+   * Each line in the text files is a new element in the resulting Dataset. For example:
```
Comment (jaceklaskowski) [inline sql/core/src/main/scala/org/apache/spark/sql/streaming/DataStreamReader.scala]: s/element/record?

File: sql/core/src/main/scala/org/apache/spark/sql/streaming/DataStreamReader.scala [hunk 3]
```diff
@@ -281,6 +281,31 @@ final class DataStreamReader private[sql](sparkSession: SparkSession) extends Lo
...
+   *   // Java:
+   *   spark.read().textFile("/path/to/spark/README.md")
+   * }}}
+   *
+   * @param path input path
+   * @since 2.0.0
+   */
+  def textFile(path: String): Dataset[String] = {
+    if (userSpecifiedSchema.nonEmpty) {
+      throw new AnalysisException("User specified schema not supported with `textFile`")
```
Comment (jaceklaskowski) [inline sql/core/src/main/scala/org/apache/spark/sql/streaming/DataStreamReader.scala]: user-specified

File: sql/core/src/main/scala/org/apache/spark/sql/streaming/DataStreamReader.scala [hunk 4]
```diff
@@ -281,6 +281,31 @@ final class DataStreamReader private[sql](sparkSession: SparkSession) extends Lo
...
+    }
+    text(path).select("value").as[String](sparkSession.implicits.newStringEncoder)
```
Comment (jaceklaskowski) [inline sql/core/src/main/scala/org/apache/spark/sql/streaming/DataStreamReader.scala]: I'm surprised that `sparkSession.implicits.newStringEncoder` is required here? Why is `sparkSession.implicits._` not imported here?

File: sql/core/src/main/scala/org/apache/spark/sql/streaming/DataStreamReader.scala [hunk 5]
```diff
@@ -281,6 +281,31 @@ final class DataStreamReader private[sql](sparkSession: SparkSession) extends Lo
...
+   * Loads text files and returns a [[Dataset]] of String. The underlying schema of the Dataset
+   * contains a single string column named "value".
+   *
+   * If the directory structure of the text files contains partitioning information, those are
+   * ignored in the resulting Dataset. To include partitioning information as columns, use `text`.
+   *
+   * Each line in the text files is a new element in the resulting Dataset. For example:
+   * {{{
+   *   // Scala:
+   *   spark.read.textFile("/path/to/spark/README.md")
```
Comment (jaceklaskowski) [inline sql/core/src/main/scala/org/apache/spark/sql/streaming/DataStreamReader.scala]: s/read/readStream?

File: sql/core/src/main/scala/org/apache/spark/sql/streaming/DataStreamReader.scala [hunk 6]
```diff
@@ -281,6 +281,31 @@ final class DataStreamReader private[sql](sparkSession: SparkSession) extends Lo
...
+   *
+   *   // Java:
+   *   spark.read().textFile("/path/to/spark/README.md")
```
Comment (jaceklaskowski) [inline sql/core/src/main/scala/org/apache/spark/sql/streaming/DataStreamReader.scala]: s/read/readStream?

File: sql/core/src/test/scala/org/apache/spark/sql/streaming/FileStreamSourceSuite.scala [hunk 7]
```diff
@@ -331,6 +331,24 @@ class FileStreamSourceSuite extends FileStreamSourceTest {
...
+    withTempDirs { case (src, tmp) =>
+      val textStream = spark.readStream.textFile(src.getCanonicalPath)
+      val filtered = textStream.filter($"value" contains "keep")
+
+      testStream(filtered)(
+        AddTextFileData("drop1\nkeep2\nkeep3", src, tmp),
+        CheckAnswer("keep2", "keep3"),
+        StopStream,
+        AddTextFileData("drop4\nkeep5\nkeep6", src, tmp),
+        StartStream(),
```
Comment (jaceklaskowski) [inline sql/core/src/test/scala/org/apache/spark/sql/streaming/FileStreamSourceSuite.scala]: Just wondering why `()` are here while not for `StopStream`?

File: sql/core/src/main/scala/org/apache/spark/sql/streaming/DataStreamReader.scala (hunk 1 above)
Comment (ScrapCodes) [inline sql/core/src/main/scala/org/apache/spark/sql/streaming/DataStreamReader.scala]: Thanks :)

File: sql/core/src/main/scala/org/apache/spark/sql/streaming/DataStreamReader.scala (hunk 2 above)
Comment (ScrapCodes) [inline sql/core/src/main/scala/org/apache/spark/sql/streaming/DataStreamReader.scala]: This is okay, I think. Not sure what others think.

File: sql/core/src/main/scala/org/apache/spark/sql/streaming/DataStreamReader.scala (hunk 6 above)
Comment (ScrapCodes) [inline sql/core/src/main/scala/org/apache/spark/sql/streaming/DataStreamReader.scala]: Correct !

File: sql/core/src/main/scala/org/apache/spark/sql/streaming/DataStreamReader.scala (hunk 1 above)
Comment (ScrapCodes) [inline sql/core/src/main/scala/org/apache/spark/sql/streaming/DataStreamReader.scala]: No actually it should be text files.

File: sql/core/src/test/scala/org/apache/spark/sql/streaming/FileStreamSourceSuite.scala (hunk 7 above)
Comment (ScrapCodes) [inline sql/core/src/test/scala/org/apache/spark/sql/streaming/FileStreamSourceSuite.scala]: This is a correct question, but it was a choice already made by others.

File: sql/core/src/main/scala/org/apache/spark/sql/streaming/DataStreamReader.scala (hunk 4 above)
Comment (ScrapCodes) [inline sql/core/src/main/scala/org/apache/spark/sql/streaming/DataStreamReader.scala]: If I remember correctly, in Spark codebase we prefer explicitly stating the implicit used.

File: sql/core/src/main/scala/org/apache/spark/sql/streaming/DataStreamReader.scala (hunk 3 above)
Comment (holdenk) [inline sql/core/src/main/scala/org/apache/spark/sql/streaming/DataStreamReader.scala]: Since this check is presumably copied from the similar function in DataFrameReader, we should probably keep the exception the same as DataFrameReader (so either update it too or leave this as is).
Also In the SQL code base we use "User specified"  24 times and "User-specified" 5 times.

**Comment by ScrapCodes** [comment]
@marmbrus Do you think this is useful ?

File: sql/core/src/test/scala/org/apache/spark/sql/streaming/FileStreamSourceSuite.scala (hunk 7 above)
Comment (marmbrus) [inline sql/core/src/test/scala/org/apache/spark/sql/streaming/FileStreamSourceSuite.scala]: Stopping has no parameters, starting you might choose to change trigger interval.

**Comment by ScrapCodes** [comment]
@tdas Ping !

File: sql/core/src/main/scala/org/apache/spark/sql/streaming/DataStreamReader.scala [hunk 8]
```diff
@@ -21,13 +21,13 @@ import scala.collection.JavaConverters._
...
 import org.apache.spark.internal.Logging
-import org.apache.spark.sql.{DataFrame, Dataset, SparkSession}
+import org.apache.spark.sql.{AnalysisException, DataFrame, Dataset, SparkSession}
 import org.apache.spark.sql.execution.datasources.DataSource
 import org.apache.spark.sql.execution.streaming.StreamingRelation
 import org.apache.spark.sql.types.StructType
 
 /**
- * Interface used to load a streaming [[Dataset]] from external storage systems (e.g. file systems,
+ * Class used to load a streaming [[Dataset]] from external storage systems (e.g. file systems,
```
Comment (marmbrus) [inline sql/core/src/main/scala/org/apache/spark/sql/streaming/DataStreamReader.scala]: This change seems unrelated and takes us out of sync with the batch version.  I don't think this means a JVM interface, but rather the `interface` in API.

File: sql/core/src/main/scala/org/apache/spark/sql/streaming/DataStreamReader.scala [hunk 9]
```diff
@@ -311,6 +311,37 @@ final class DataStreamReader private[sql](sparkSession: SparkSession) extends Lo
   @Experimental
   def text(path: String): DataFrame = format("text").load(path)
 
+  /**
+   * Loads text file(s) and returns a [[Dataset]] of String. The underlying schema of the Dataset
```
Comment (jodersky) [inline sql/core/src/main/scala/org/apache/spark/sql/streaming/DataStreamReader.scala]: Should text files be plural here? The api would be more intuitive by copying the non-streaming equivalent with a vararg-method for multiple parameters

File: sql/core/src/main/scala/org/apache/spark/sql/streaming/DataStreamReader.scala (hunk 8 above)
Comment (ScrapCodes) [inline sql/core/src/main/scala/org/apache/spark/sql/streaming/DataStreamReader.scala]: Understood, thanks for correcting !

File: sql/core/src/main/scala/org/apache/spark/sql/streaming/DataStreamReader.scala (hunk 9 above)
Comment (ScrapCodes) [inline sql/core/src/main/scala/org/apache/spark/sql/streaming/DataStreamReader.scala]: I would like to be corrected, as I just followed the convention over here. Since this class does not have any vararg method for other APIs, I was doubtful in adding one myself.

File: sql/core/src/test/scala/org/apache/spark/sql/streaming/FileStreamSourceSuite.scala [hunk 10]
```diff
@@ -378,6 +378,24 @@ class FileStreamSourceSuite extends FileStreamSourceTest {
     }
   }
 
+  test("read from textfile") {
+    withTempDirs { case (src, tmp) =>
+      val textStream = spark.readStream.textFile(src.getCanonicalPath)
+      val filtered = textStream.filter($"value" contains "keep")
```
Comment (marmbrus) [inline sql/core/src/test/scala/org/apache/spark/sql/streaming/FileStreamSourceSuite.scala]: One last comment.  I'd use the typed API here since that is the whole point of `textFile` vs `text`.

File: sql/core/src/main/scala/org/apache/spark/sql/streaming/DataStreamReader.scala (hunk 9 above)
Comment (marmbrus) [inline sql/core/src/main/scala/org/apache/spark/sql/streaming/DataStreamReader.scala]: It might be weird to add var args, since the streaming case would always be to watch a directory (not list a bunch of files).  I think its fine to leave it out for now.

This is existing, but its a little odd that the methods in this file talk about `loading files` rather than `watching directories of files and processing them as they appear`.

File: sql/core/src/test/scala/org/apache/spark/sql/streaming/FileStreamSourceSuite.scala (hunk 10 above)
Comment (ScrapCodes) [inline sql/core/src/test/scala/org/apache/spark/sql/streaming/FileStreamSourceSuite.scala]: updated it. :)

**Comment by marmbrus** [comment]
Thanks, merging to master.
